components/butterfly/src/rumor.rs

```diff
@@ -408,17 +498,50 @@ impl<T> RumorStore<T> where T: Rumor
     /// We don't care if this repeats - it just needs to be unique for any given two states, which
     /// it will be.
     fn increment_update_counter(&self) { self.update_counter.fetch_add(1, Ordering::Relaxed); }
+
+    /// Find rumors in our rumor store that have expired.
+    fn partitioned_rumors(&self, expiration_date: DateTime<Utc>) -> (Vec<T>, Vec<T>) {
```

Comment:
I'll investigate this further and see how it goes.

Reply:
It's possible that I'm misunderstanding [the docs for partition](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition) but it sounds like that function consumes the iterator and creates 2 collections.  If we want to keep the iterator around to avoid unnecessary allocations, then maybe `partition` isn't the right thing to use - and we instead just directly retrieve the rumors that we want in the `expired()` and `live()` functions.

Reply:
Alternatively, it looks like if I change the return type of `partitioned_rumors` to match the return type of `partition` itself (instead of `Vec`), it still compiles.

```rust
    fn partitioned<B>(&self, expiration_date: DateTime<Utc>) -> (B, B)
        where B: Default + Extend<T>
    {
        self.read_entries()
            .values()
            .flat_map(HashMap::values)
            .cloned()
            .partition(|rumor| rumor.expiration().expired(expiration_date))
    }
```

Is that sufficient to avoid the allocations?

Reply:
I've committed the above change, but I'm not entirely certain how to tell if it accomplishes our goal here or not.  We can revisit this if necessary.

Reply:
I don't think this avoids allocations. Looking at [the implementation of `std::iter::Iterator::partition`](https://github.com/rust-lang/rust/blob/e13fe7ff5e482b626714ac6ec41b8d44e7c22ed1/src/libcore/iter/traits/iterator.rs#L1493), it creates two new collections and calls [`extend`](http://doc.rust-lang.org/1.36.0/std/iter/trait.Extend.html) on them. That method is going to have to allocate. And even before we get to that point, the `cloned` adaptor is going to clone every element.

In the end, we may be unable to avoid some cloning, but we can almost certainly limit it to the half of the partition that we care about returning.

There's a lot of stuff still in flux with this PR, so maybe we can wait for some of the other issues to settle first, but if at that point you want to work more closely to see how we can make this more efficient, I'd be happy to collaborate.